src/types.ts

```diff
@@ -311,7 +310,7 @@ export type TransitionsConfig<TContext, TEvent extends EventObject> = {
     | StateNode<TContext, any, TEvent>
     | TransitionConfig<
         TContext,
-        TEvent extends { type: K } ? TEvent : EventObject
+        K extends TEvent['type'] ? Extract<TEvent, { type: K }> : EventObject
```

Comment:
This improves inference by propagating correctly narrowed down type.

Before:
<img width="341" alt="Screenshot 2019-08-24 at 21 21 24" src="https://user-images.githubusercontent.com/9800850/63641849-3dd10700-c6b5-11e9-85be-c4462acfd4fa.png">
After:
<img width="433" alt="Screenshot 2019-08-24 at 21 22 56" src="https://user-images.githubusercontent.com/9800850/63641860-5f31f300-c6b5-11e9-8104-528da7c4a9dd.png">